core/storage/user/gae_models.py

```diff
@@ -1959,61 +1959,130 @@ def _get_instance_id(cls, user_id, score_category):
+                ' given id: %s' % instance_id
+            )
 
-        cls(
+        user_scoring_model = cls(
             id=instance_id, user_id=user_id, score_category=score_category,
-            score=score).put()
+            score=score)
+        user_scoring_model.put()
+        return user_scoring_model
 
     @classmethod
-    def increment_score_for_user(cls, user_id, score_category, increment_by):
-        """Increment the score of the user in the category by the given amount.
+    def create_multi(cls, user_score_identifiers, score):
+        """Creates new UserContributionScoringModel entries.
 
         Args:
-            user_id: str. The id of the user.
-            score_category: str. The category of the suggestion.
-            increment_by: float. The amount to increase the score of the user
-                by. May be negative, in which case the score is reduced.
+            user_score_identifiers: list(FullyQualifiedUserScoreIdentifier). A
+                list of unique (user_id, score_category) pairs that correspond
+                to the id of the users and the suggestion score categories to
+                create the scoring models for.
+            score: float. The score of the users.
+
+        Returns:
+            list(UserContributionScoringModel). The user scoring models that
+            were created.
+
+        Raises:
+            Exception. There is already an entry with the given id.
         """
-        instance_id = cls._get_instance_id(user_id, score_category)
-        model = cls.get_by_id(instance_id)
-        if not model:
-            cls.create(user_id, score_category, increment_by)
-        else:
-            model.score += increment_by
-            model.put()
+        user_ids = [
+            user_score_identifier.user_id for user_score_identifier in
+            user_score_identifiers
+        ]
+        score_categories = [
+            user_score_identifier.score_category for user_score_identifier in
+            user_score_identifiers
+        ]
+        instance_ids = [
+            cls._get_instance_id(user_id, score_category)
+            for user_id, score_category in python_utils.ZIP(
+                user_ids, score_categories)
+        ]
+        user_scoring_models = cls.get_multi(instance_ids)
+        for index, user_scoring_model in enumerate(user_scoring_models):
+            if user_scoring_model:
+                raise Exception(
+                    'There is already a UserContributionScoringModel entry with'
+                    ' the given id: %s' % instance_ids[index]
+                )
+
+        new_user_scoring_models = [
+            cls(
+                id=instance_id, user_id=user_id, score_category=score_category,
+                score=score
+            )
+            for instance_id, user_id, score_category in python_utils.ZIP(
+                instance_ids, user_ids, score_categories)
+        ]
+        cls.put_multi(new_user_scoring_models)
```

Comment:
Does this need to happen transactionally to avoid races against other users changing the suggestions at the same time? Multi operations can be quite expensive and take some time.